packages/run-protocol/src/vaultFactory/vaultManager.js

```diff
@@ -151,48 +204,122 @@ export const makeVaultManager = (
+      liquidateAndRemove,
     );
-    return Promise.all(promises);
+    await Promise.all(toLiquidate);
   };
 
+  /**
+   *
+   * @param {bigint} updateTime
+   * @param {ZCFSeat} poolIncrementSeat
+   */
   const chargeAllVaults = async (updateTime, poolIncrementSeat) => {
-    const poolIncrement = sortedVaultKits.reduce(
-      (total, vaultPair) =>
-        AmountMath.add(
-          total,
-          vaultPair.vaultKit.accrueInterestAndAddToPool(updateTime),
-        ),
-      AmountMath.makeEmpty(runBrand),
+    trace('chargeAllVault', { updateTime });
+    const interestCalculator = makeInterestCalculator(
+      shared.getInterestRate(),
+      shared.getChargingPeriod(),
+      shared.getRecordingPeriod(),
+    );
+
+    // calculate delta of accrued debt
+    const debtStatus = interestCalculator.calculateReportingPeriod(
+      {
+        latestInterestUpdate,
+        newDebt: totalDebt,
+        interest: 0n, // XXX this is always zero, doesn't need to be an option
+      },
+      updateTime,
+    );
+    const interestAccrued = debtStatus.interest;
+
+    // done if none
+    if (interestAccrued === 0n) {
+      return;
+    }
+
+    // NB: This method of inferring the compounded rate from the ratio of debts
+    // acrrued suffers slightly from the integer nature of debts. However in
+    // testing with small numbers there's 5 digits of precision, and with large
+    // numbers the ratios tend towards ample precision. Because this calculation
+    // is over all debts of the vault the numbers will be reliably large.
+    compoundedInterest = calculateCompoundedInterest(
+      compoundedInterest,
+      totalDebt,
+      debtStatus.newDebt,
     );
-    sortedVaultKits.updateAllDebts();
+    totalDebt += interestAccrued;
+
+    // mint that much RUN for the reward pool
+    const rewarded = AmountMath.make(runBrand, interestAccrued);
+    runMint.mintGains(harden({ RUN: rewarded }), poolIncrementSeat);
+    reallocateReward(rewarded, poolIncrementSeat);
+
+    // update running tally of total debt against this collateral
+    ({ latestInterestUpdate } = debtStatus);
+
+    const payload = harden({
+      compoundedInterest,
+      latestInterestUpdate,
+      totalDebt: AmountMath.make(runBrand, totalDebt),
+    });
+    updater.updateState(payload);
+
+    trace('chargeAllVaults complete', payload);
+
     reschedulePriceCheck();
-    runMint.mintGains(harden({ RUN: poolIncrement }), poolIncrementSeat);
-    reallocateReward(poolIncrement, poolIncrementSeat);
+  };
+
+  /**
+   * @param {Amount<NatValue>} oldDebtOnVault
+   * @param {Amount<NatValue>} newDebtOnVault
+   */
+  const applyDebtDelta = (oldDebtOnVault, newDebtOnVault) => {
+    const delta = newDebtOnVault.value - oldDebtOnVault.value;
+    trace(`updating total debt ${totalDebt} by ${delta}`);
+    if (delta === 0n) {
+      // nothing to do
+      return;
+    }
+
+    if (delta > 0n) {
+      // add the amount
+      totalDebt += delta;
```

Comment:
Why isn't totalDebt an Amount?  There's no reason not to have it be a checked value where our units math is doing the right thing.

Reply:
The bug fixed in in d2560baf790b76891a039d132ece224f52e88824 was largely due to wrapping and unwrapping amount values for totalDebt. So I'm convinced you're right that it's better to keep it as an amount: e71db04c294b0908b16d1752b71a5f641f55a4cd